File: Class 2/src/graphwidget.cpp
#include "graphwidget.h"

#include <QLabel>
#include <QPaintEvent>
#include <QPainter>
#include <QPainterPath>
#include <QPen>
#include <QResizeEvent>
#include <QSignalBlocker>
#include <QSlider>
#include <QtGlobal>
#include <algorithm>
#include <cmath>

namespace {
constexpr int leftMargin = 52;
constexpr int rightMargin = 20;
constexpr int topMargin = 20;
constexpr int bottomMargin = 56;
constexpr int gridLines = 8;

QColor colorForIndex(int index)
{
    static const QVector<QColor> palette = {
        QColor(84, 191, 160),
        QColor(235, 125, 89),
        QColor(106, 160, 255),
        QColor(232, 193, 112),
        QColor(192, 132, 252),
        QColor(123, 203, 111)
    };
    return palette.at(index % palette.size());
}
}

GraphWidget::GraphWidget(QWidget *parent)
    : QWidget(parent),
      timeLabel_(new QLabel(this)),
      timeSlider_(new QSlider(Qt::Horizontal, this))
{
    setMinimumSize(420, 320);
    setAutoFillBackground(true);

    timeSlider_->setRange(0, 200);
    timeSlider_->setValue(0);
    timeLabel_->setAlignment(Qt::AlignRight | Qt::AlignVCenter);
    updateTimeLabel();

    connect(timeSlider_, &QSlider::valueChanged, this, [this]() {
        updateTimeLabel();
        recomputeCurves();
    });
}

void GraphWidget::setPlotDefinition(const NamedFunction &function, double xMin, double xMax, double tf, const QList<int> &nValues)
{
    selectedFunction_ = function;
    xMin_ = xMin;
    xMax_ = xMax;
    tf_ = tf;

    timeSlider_->setEnabled(tf_ > 0.0);
    {
        const QSignalBlocker blocker(timeSlider_);
        timeSlider_->setValue(0);
    }
    updateTimeLabel();

    curves_.clear();
    errorMessage_.clear();

    if (!selectedFunction_.evaluator) {
        errorMessage_ = "Selecione uma funcao valida.";
        update();
        return;
    }

    nValues_.clear();
    nValues_ = nValues;
    customZoom_ = false;
    recomputeCurves();
}

void GraphWidget::resizeEvent(QResizeEvent *event)
{
    QWidget::resizeEvent(event);

    const int margin = 16;
    const int sliderHeight = 24;
    const int labelWidth = 120;
    const int bottomY = height() - margin - sliderHeight;

    timeSlider_->setGeometry(margin, bottomY, width() - (3 * margin) - labelWidth, sliderHeight);
    timeLabel_->setGeometry(width() - margin - labelWidth, bottomY, labelWidth, sliderHeight);
}

void GraphWidget::recomputeCurves()
{
    curves_.clear();
    errorMessage_.clear();

    if (!selectedFunction_.evaluator) {
        update();
        return;
    }

    constexpr int sampleCount = 400;
    const double t = currentTime();

    for (int index = 0; index < nValues_.size(); ++index) {
        const int n = nValues_.at(index);
        CurveData curve;
        curve.n = n;
        curve.color = colorForIndex(index);
        curve.samples.reserve(sampleCount);

        for (int sample = 0; sample < sampleCount; ++sample) {
            const double ratio = sampleCount > 1 ? static_cast<double>(sample) / (sampleCount - 1) : 0.0;
            const double x = xMin_ + (xMax_ - xMin_) * ratio;

            const double y = selectedFunction_.evaluator(x, t, n);
            if (!std::isfinite(y)) {
                continue;
            }

            curve.samples.append(QPointF(x, y));
        }

        if (!curve.samples.isEmpty()) {
            curves_.append(curve);
        }
    }

    if (curves_.isEmpty()) {
        errorMessage_ = "Nenhum ponto valido foi gerado para o grafico.";
    }

    resetZoom();

    update();
}

void GraphWidget::updateTimeLabel()
{
    timeLabel_->setText(QString("t = %1 / %2").arg(currentTime(), 0, 'f', 3).arg(tf_, 0, 'f', 3));
}

double GraphWidget::currentTime() const
{
    if (tf_ <= 0.0) {
        return 0.0;
    }

    return tf_ * static_cast<double>(timeSlider_->value()) / static_cast<double>(timeSlider_->maximum());
}

QRectF GraphWidget::plotArea() const
{
    return QRectF(
        leftMargin,
        topMargin,
        std::max(120, width() - leftMargin - rightMargin),
        std::max(120, height() - topMargin - bottomMargin)
    );
}

QPointF GraphWidget::mapToScreen(const QPointF &point, const QRectF &area, double xMin, double xMax, double yMin, double yMax) const
{
    const double xRatio = (point.x() - xMin) / (xMax - xMin);
    const double yRatio = (point.y() - yMin) / (yMax - yMin);

    return QPointF(
        area.left() + xRatio * area.width(),
        area.bottom() - yRatio * area.height()
    );
}

double GraphWidget::mapToModelX(double xScreen, const QRectF &area, double xMin, double xMax) const
{
    const double ratio = (xScreen - area.left()) / area.width();
    return xMin + ratio * (xMax - xMin);
}

void GraphWidget::resetZoom()
{
    visibleXMin_ = xMin_;
    visibleXMax_ = xMax_;

    if (curves_.isEmpty()) {
        visibleYMin_ = -1.0;
        visibleYMax_ = 1.0;
        customZoom_ = false;
        return;
    }

    double yMin = curves_.first().samples.first().y();
    double yMax = yMin;

    for (const CurveData &curve : curves_) {
        for (const QPointF &point : curve.samples) {
            yMin = std::min(yMin, point.y());
            yMax = std::max(yMax, point.y());
        }
    }

    if (std::abs(yMax - yMin) < 1e-9) {
        yMin -= 1.0;
        yMax += 1.0;
    } else {
        const double margin = (yMax - yMin) * 0.15;
        yMin -= margin;
        yMax += margin;
    }

    visibleYMin_ = yMin;
    visibleYMax_ = yMax;
    customZoom_ = false;
}

void GraphWidget::wheelEvent(QWheelEvent *event)
{
    if (curves_.isEmpty() || !errorMessage_.isEmpty()) {
        event->ignore();
        return;
    }

    const QRectF area = plotArea();
    if (!area.contains(event->position())) {
        event->ignore();
        return;
    }

    const double xMin = visibleXMin_;
    const double xMax = visibleXMax_;
    const double yMin = visibleYMin_;
    const double yMax = visibleYMax_;
    const double factor = event->angleDelta().y() > 0 ? 0.85 : 1.15;
    const double centerX = mapToModelX(event->position().x(), area, xMin, xMax);
    const double centerY = yMax - ((event->position().y() - area.top()) / area.height()) * (yMax - yMin);
    const double newWidth = std::max(0.001, (xMax - xMin) * factor);
    const double newHeight = std::max(0.001, (yMax - yMin) * factor);
    const double ratioX = (centerX - xMin) / (xMax - xMin);
    const double ratioY = (centerY - yMin) / (yMax - yMin);

    visibleXMin_ = centerX - ratioX * newWidth;
    visibleXMax_ = visibleXMin_ + newWidth;
    visibleYMin_ = centerY - ratioY * newHeight;
    visibleYMax_ = visibleYMin_ + newHeight;
    customZoom_ = true;

    update();
    event->accept();
}

void GraphWidget::mouseDoubleClickEvent(QMouseEvent *event)
{
    if (event->button() == Qt::LeftButton) {
        resetZoom();
        update();
        event->accept();
        return;
    }

    QWidget::mouseDoubleClickEvent(event);
}

void GraphWidget::paintEvent(QPaintEvent *event)
{
    QWidget::paintEvent(event);

    QPainter painter(this);
    painter.setRenderHint(QPainter::Antialiasing, true);
    painter.fillRect(rect(), QColor(20, 24, 31));

    const QRectF plotRect = plotArea();
    painter.setPen(QColor(58, 66, 84));
    painter.drawRoundedRect(plotRect.adjusted(-1, -1, 1, 1), 10, 10);

    if (!errorMessage_.isEmpty()) {
        painter.setPen(QColor(226, 149, 120));
        painter.drawText(plotRect, Qt::AlignCenter | Qt::TextWordWrap, errorMessage_);
        return;
    }

    if (curves_.isEmpty()) {
        painter.setPen(QColor(160, 171, 190));
        painter.drawText(plotRect, Qt::AlignCenter, "Aguardando parametros para plotagem.");
        return;
    }

    const double xMin = visibleXMin_;
    const double xMax = visibleXMax_;
    const double yMin = visibleYMin_;
    const double yMax = visibleYMax_;

    QPen gridPen(QColor(44, 50, 63));
    gridPen.setWidth(1);
    painter.setPen(gridPen);

    for (int index = 0; index <= gridLines; ++index) {
        const double ratio = static_cast<double>(index) / gridLines;
        const double x = plotRect.left() + ratio * plotRect.width();
        const double y = plotRect.top() + ratio * plotRect.height();

        painter.drawLine(QPointF(x, plotRect.top()), QPointF(x, plotRect.bottom()));
        painter.drawLine(QPointF(plotRect.left(), y), QPointF(plotRect.right(), y));

        const double xValue = xMin + ratio * (xMax - xMin);
        const double yValue = yMax - ratio * (yMax - yMin);

        painter.setPen(QColor(130, 141, 160));
        painter.drawText(QRectF(x - 28, plotRect.bottom() + 6, 56, 16), Qt::AlignCenter, QString::number(xValue, 'g', 4));
        painter.drawText(QRectF(4, y - 8, leftMargin - 10, 16), Qt::AlignRight | Qt::AlignVCenter, QString::number(yValue, 'g', 4));
        painter.setPen(gridPen);
    }

    QPen axisPen(QColor(115, 132, 158));
    axisPen.setWidth(2);
    painter.setPen(axisPen);

    if (xMin <= 0.0 && xMax >= 0.0) {
        painter.drawLine(
            mapToScreen(QPointF(0.0, yMax), plotRect, xMin, xMax, yMin, yMax),
            mapToScreen(QPointF(0.0, yMin), plotRect, xMin, xMax, yMin, yMax)
        );
    }

    if (yMin <= 0.0 && yMax >= 0.0) {
        painter.drawLine(
            mapToScreen(QPointF(xMin, 0.0), plotRect, xMin, xMax, yMin, yMax),
            mapToScreen(QPointF(xMax, 0.0), plotRect, xMin, xMax, yMin, yMax)
        );
    }

    for (const CurveData &curve : curves_) {
        QPainterPath path;
        bool firstPoint = true;

        for (const QPointF &sample : curve.samples) {
            if (sample.x() < xMin || sample.x() > xMax) {
                firstPoint = true;
                continue;
            }
            const QPointF mappedPoint = mapToScreen(sample, plotRect, xMin, xMax, yMin, yMax);
            if (firstPoint) {
                path.moveTo(mappedPoint);
                firstPoint = false;
            } else {
                path.lineTo(mappedPoint);
            }
        }

        const bool isPrimaryCurve = !nValues_.isEmpty() && curve.n == nValues_.front();
        painter.setPen(QPen(curve.color, isPrimaryCurve ? 3.0 : 2.0));
        painter.drawPath(path);
    }

    const QRectF legendBox(plotRect.right() - 150, plotRect.top() + 10, 140, 28 + curves_.size() * 22);
    painter.setPen(QColor(58, 66, 84));
    painter.setBrush(QColor(16, 20, 26, 210));
    painter.drawRoundedRect(legendBox, 8, 8);

    int legendY = static_cast<int>(legendBox.top()) + 14;
    for (const CurveData &curve : curves_) {
        painter.setPen(QPen(curve.color, 2));
        painter.drawLine(QPointF(legendBox.left() + 10, legendY), QPointF(legendBox.left() + 28, legendY));
        painter.setPen(QColor(230, 235, 242));
        painter.drawText(QRectF(legendBox.left() + 34, legendY - 9, legendBox.width() - 40, 18), Qt::AlignLeft | Qt::AlignVCenter, QString("N = %1").arg(curve.n));
        legendY += 22;
    }

    painter.setPen(QColor(214, 220, 230));
    painter.drawText(QRectF(plotRect.left(), 0, plotRect.width(), 18), Qt::AlignCenter,
                     QString("u(x,t,N) via Phi   |   t = %1").arg(currentTime(), 0, 'f', 3));
}

File: Class 2 - versao 2/src/class2_v2_app/graph_widget.py
from __future__ import annotations

import math
from dataclasses import dataclass, field

from PyQt6.QtCore import QPointF, QRectF, Qt
from PyQt6.QtGui import QColor, QMouseEvent, QPaintEvent, QPainter, QPainterPath, QPen, QWheelEvent
from PyQt6.QtWidgets import QWidget

from .function_catalog import NamedFunction
from .time_method_catalog import TimeMethod, advance_curve, evaluate_exact_curve

LEFT_MARGIN = 52
RIGHT_MARGIN = 20
TOP_MARGIN = 20
BOTTOM_MARGIN = 36
GRID_LINES = 8
SAMPLE_COUNT = 400


@dataclass
class CurveData:
    label: str
    samples: list[QPointF] = field(default_factory=list)
    color: QColor = field(default_factory=lambda: QColor(84, 191, 160))
    width: float = 2.0
    pen_style: Qt.PenStyle = Qt.PenStyle.SolidLine


def color_for_index(index: int) -> QColor:
    palette = [
        QColor(84, 191, 160),
        QColor(235, 125, 89),
        QColor(106, 160, 255),
        QColor(232, 193, 112),
        QColor(192, 132, 252),
        QColor(123, 203, 111),
    ]
    return palette[index % len(palette)]


def method_style(method_id: str) -> tuple[Qt.PenStyle, float]:
    styles = {
        "forward": (Qt.PenStyle.DashDotLine, 2.0),
        "backward": (Qt.PenStyle.DotLine, 2.0),
        "central": (Qt.PenStyle.DashLine, 2.4),
    }
    return styles.get(method_id, (Qt.PenStyle.SolidLine, 2.0))


class GraphWidget(QWidget):
    def __init__(self, parent: QWidget | None = None) -> None:
        super().__init__(parent)

        self.curves: list[CurveData] = []
        self.error_message = ""
        self.x_min = 0.0
        self.x_max = 1.0
        self.tf = 0.0
        self.n_value = 1
        self.time_steps_values: list[int] = []
        self.visible_x_min = 0.0
        self.visible_x_max = 1.0
        self.visible_y_min = -1.0
        self.visible_y_max = 1.0
        self.selected_function: NamedFunction | None = None
        self.selected_methods: list[TimeMethod] = []

        self.setMinimumSize(420, 320)
        self.setAutoFillBackground(True)

    def set_plot_definition(
        self,
        function: NamedFunction,
        x_min: float,
        x_max: float,
        tf: float,
        n_value: int,
        time_steps_values: list[int],
        selected_methods: list[TimeMethod],
    ) -> None:
        self.selected_function = function
        self.x_min = x_min
        self.x_max = x_max
        self.tf = tf
        self.n_value = n_value
        self.time_steps_values = list(time_steps_values)
        self.selected_methods = list(selected_methods)
        self.error_message = ""
        self.curves.clear()

        if self.selected_function is None:
            self.error_message = "Selecione uma funcao valida."
            self.update()
            return

        self.recompute_curves()

    def recompute_curves(self) -> None:
        self.curves.clear()
        self.error_message = ""

        if self.selected_function is None:
            self.update()
            return

        x_values = [
            self.x_min + (self.x_max - self.x_min) * (sample / (SAMPLE_COUNT - 1))
            for sample in range(SAMPLE_COUNT)
        ]

        initial_curve = evaluate_exact_curve(self.selected_function.evaluator, x_values, 0.0, self.n_value)
        self._append_curve(
            label=f"t = 0 | N = {self.n_value}",
            x_values=x_values,
            y_values=initial_curve,
            color=QColor(190, 198, 212),
            pen_style=Qt.PenStyle.DashLine,
            width=1.8,
        )

        final_exact_curve = evaluate_exact_curve(self.selected_function.evaluator, x_values, self.tf, self.n_value)
        self._append_curve(
            label=f"Exata tf | N = {self.n_value}",
            x_values=x_values,
            y_values=final_exact_curve,
            color=QColor(255, 255, 255),
            pen_style=Qt.PenStyle.SolidLine,
            width=3.0,
        )

        for nt_index, nt_value in enumerate(self.time_steps_values):
            nt_color = color_for_index(nt_index)
            for method in self.selected_methods:
                method_curve = advance_curve(
                    self.selected_function.evaluator,
                    x_values,
                    self.tf,
                    nt_value,
                    self.n_value,
                    method.identifier,
                )
                pen_style, width = method_style(method.identifier)
                self._append_curve(
                    label=f"{method.label} | Nt = {nt_value}",
                    x_values=x_values,
                    y_values=method_curve,
                    color=nt_color,
                    pen_style=pen_style,
                    width=width,
                )

        if not self.curves:
            self.error_message = "Nenhuma curva valida foi gerada para o grafico."

        self.reset_zoom()
        self.update()

    def _append_curve(
        self,
        label: str,
        x_values: list[float],
        y_values: list[float],
        color: QColor,
        pen_style: Qt.PenStyle,
        width: float,
    ) -> None:
        curve = CurveData(label=label, color=color, pen_style=pen_style, width=width)
        for x_value, y_value in zip(x_values, y_values):
            if math.isfinite(y_value):
                curve.samples.append(QPointF(x_value, y_value))
        if curve.samples:
            self.curves.append(curve)

    def plot_area(self) -> QRectF:
        return QRectF(
            LEFT_MARGIN,
            TOP_MARGIN,
            max(120, self.width() - LEFT_MARGIN - RIGHT_MARGIN),
            max(120, self.height() - TOP_MARGIN - BOTTOM_MARGIN),
        )

    def map_to_screen(self, point: QPointF, area: QRectF, x_min: float, x_max: float, y_min: float, y_max: float) -> QPointF:
        x_ratio = (point.x() - x_min) / (x_max - x_min)
        y_ratio = (point.y() - y_min) / (y_max - y_min)
        return QPointF(area.left() + x_ratio * area.width(), area.bottom() - y_ratio * area.height())

    def map_to_model_x(self, x_screen: float, area: QRectF, x_min: float, x_max: float) -> float:
        ratio = (x_screen - area.left()) / area.width()
        return x_min + ratio * (x_max - x_min)

    def reset_zoom(self) -> None:
        self.visible_x_min = self.x_min
        self.visible_x_max = self.x_max

        if not self.curves:
            self.visible_y_min = -1.0
            self.visible_y_max = 1.0
            return

        y_values = [point.y() for curve in self.curves for point in curve.samples]
        y_min = min(y_values)
        y_max = max(y_values)

        if abs(y_max - y_min) < 1e-9:
            y_min -= 1.0
            y_max += 1.0
        else:
            margin = (y_max - y_min) * 0.15
            y_min -= margin
            y_max += margin

        self.visible_y_min = y_min
        self.visible_y_max = y_max

    def wheelEvent(self, event: QWheelEvent) -> None:  # noqa: N802
        if not self.curves or self.error_message:
            event.ignore()
            return
        area = self.plot_area()
        if not area.contains(event.position()):
            event.ignore()
            return
        x_min = self.visible_x_min
        x_max = self.visible_x_max
        y_min = self.visible_y_min
        y_max = self.visible_y_max
        factor = 0.85 if event.angleDelta().y() > 0 else 1.15
        center_x = self.map_to_model_x(event.position().x(), area, x_min, x_max)
        center_y = y_max - ((event.position().y() - area.top()) / area.height()) * (y_max - y_min)
        new_width = max(0.001, (x_max - x_min) * factor)
        new_height = max(0.001, (y_max - y_min) * factor)
        ratio_x = (center_x - x_min) / (x_max - x_min)
        ratio_y = (center_y - y_min) / (y_max - y_min)
        self.visible_x_min = center_x - ratio_x * new_width
        self.visible_x_max = self.visible_x_min + new_width
        self.visible_y_min = center_y - ratio_y * new_height
        self.visible_y_max = self.visible_y_min + new_height
        self.update()
        event.accept()

    def mouseDoubleClickEvent(self, event: QMouseEvent) -> None:  # noqa: N802
        if event.button() == Qt.MouseButton.LeftButton:
            self.reset_zoom()
            self.update()
            event.accept()
            return
        super().mouseDoubleClickEvent(event)

    def paintEvent(self, event: QPaintEvent) -> None:  # noqa: N802
        super().paintEvent(event)
        painter = QPainter(self)
        painter.setRenderHint(QPainter.RenderHint.Antialiasing, True)
        painter.fillRect(self.rect(), QColor(20, 24, 31))
        plot_rect = self.plot_area()
        painter.setPen(QColor(58, 66, 84))
        painter.drawRoundedRect(plot_rect.adjusted(-1.0, -1.0, 1.0, 1.0), 10.0, 10.0)
        if self.error_message:
            painter.setPen(QColor(226, 149, 120))
            painter.drawText(plot_rect, int(Qt.AlignmentFlag.AlignCenter | Qt.AlignmentFlag.TextWordWrap), self.error_message)
            return
        if not self.curves:
            painter.setPen(QColor(160, 171, 190))
            painter.drawText(plot_rect, int(Qt.AlignmentFlag.AlignCenter), "Aguardando parametros para plotagem.")
            return
        x_min = self.visible_x_min
        x_max = self.visible_x_max
        y_min = self.visible_y_min
        y_max = self.visible_y_max
        grid_pen = QPen(QColor(44, 50, 63))
        grid_pen.setWidth(1)
        painter.setPen(grid_pen)
        for index in range(GRID_LINES + 1):
            ratio = index / GRID_LINES
            x = plot_rect.left() + ratio * plot_rect.width()
            y = plot_rect.top() + ratio * plot_rect.height()
            painter.drawLine(QPointF(x, plot_rect.top()), QPointF(x, plot_rect.bottom()))
            painter.drawLine(QPointF(plot_rect.left(), y), QPointF(plot_rect.right(), y))
            x_value = x_min + ratio * (x_max - x_min)
            y_value = y_max - ratio * (y_max - y_min)
            painter.setPen(QColor(130, 141, 160))
            painter.drawText(QRectF(x - 28, plot_rect.bottom() + 6, 56, 16), int(Qt.AlignmentFlag.AlignCenter), f"{x_value:.4g}")
            painter.drawText(QRectF(4, y - 8, LEFT_MARGIN - 10, 16), int(Qt.AlignmentFlag.AlignRight | Qt.AlignmentFlag.AlignVCenter), f"{y_value:.4g}")
            painter.setPen(grid_pen)
        axis_pen = QPen(QColor(115, 132, 158))
        axis_pen.setWidth(2)
        painter.setPen(axis_pen)
        if x_min <= 0.0 <= x_max:
            painter.drawLine(self.map_to_screen(QPointF(0.0, y_max), plot_rect, x_min, x_max, y_min, y_max), self.map_to_screen(QPointF(0.0, y_min), plot_rect, x_min, x_max, y_min, y_max))
        if y_min <= 0.0 <= y_max:
            painter.drawLine(self.map_to_screen(QPointF(x_min, 0.0), plot_rect, x_min, x_max, y_min, y_max), self.map_to_screen(QPointF(x_max, 0.0), plot_rect, x_min, x_max, y_min, y_max))
        for curve in self.curves:
            path = QPainterPath()
            first_point = True
            for sample in curve.samples:
                if sample.x() < x_min or sample.x() > x_max:
                    first_point = True
                    continue
                mapped = self.map_to_screen(sample, plot_rect, x_min, x_max, y_min, y_max)
                if first_point:
                    path.moveTo(mapped)
                    first_point = False
                else:
                    path.lineTo(mapped)
            pen = QPen(curve.color, curve.width)
            pen.setStyle(curve.pen_style)
            painter.setPen(pen)
            painter.drawPath(path)
        legend_height = 28 + len(self.curves) * 22
        legend_box = QRectF(plot_rect.right() - 290, plot_rect.top() + 10, 280, legend_height)
        painter.setPen(QColor(58, 66, 84))
        painter.setBrush(QColor(16, 20, 26, 215))
        painter.drawRoundedRect(legend_box, 8.0, 8.0)
        legend_y = int(legend_box.top()) + 14
        for curve in self.curves:
            pen = QPen(curve.color, 2)
            pen.setStyle(curve.pen_style)
            painter.setPen(pen)
            painter.drawLine(QPointF(legend_box.left() + 10, legend_y), QPointF(legend_box.left() + 30, legend_y))
            painter.setPen(QColor(230, 235, 242))
            painter.drawText(QRectF(legend_box.left() + 36, legend_y - 9, legend_box.width() - 42, 18), int(Qt.AlignmentFlag.AlignLeft | Qt.AlignmentFlag.AlignVCenter), curve.label)
            legend_y += 22
        painter.setPen(QColor(214, 220, 230))
        painter.drawText(QRectF(plot_rect.left(), 0, plot_rect.width(), 18), int(Qt.AlignmentFlag.AlignCenter), f"Comparacao temporal | t = 0, tf = {self.tf:.3f}, N = {self.n_value}")
